python/pyarrow/orc.py

```diff
@@ -117,21 +185,92 @@ def read(self, columns=None):
+compression_strategy : string, default 'speed'
+    Specify the compression strategy i.e. speed vs size reduction.
+    Valid values: {'SPEED', 'COMPRESSION'}
+row_index_stride : int, default 10000
+    Specify the row index stride i.e. the number of rows per
+    an entry in the row index.
+padding_tolerance : double, default 0.0
+    Set the padding tolerance.
+dictionary_key_size_threshold : double, default 0.0
+    Set the dictionary key size threshold. 0 to disable dictionary encoding.
+    1 to always enable dictionary encoding.
+bloom_filter_columns : None, set-like or list-like, default None
+    Set columns that use the bloom filter.
+bloom_filter_fpp: double, default 0.05
+    Set false positive probability of the bloom filter.
+"""
+
+
 class ORCWriter:
-    """
-    Writer interface for a single ORC file
+    __doc__ = """
+Writer interface for a single ORC file
 
-    Parameters
-    ----------
-    where : str or pyarrow.io.NativeFile
-        Writable target. For passing Python file objects or byte buffers,
-        see pyarrow.io.PythonFileInterface, pyarrow.io.BufferOutputStream
-        or pyarrow.io.FixedSizeBufferWriter.
-    """
+Parameters
+----------
+where : str or pyarrow.io.NativeFile
+    Writable target. For passing Python file objects or byte buffers,
+    see pyarrow.io.PythonFileInterface, pyarrow.io.BufferOutputStream
+    or pyarrow.io.FixedSizeBufferWriter.
+{}
+""".format(_orc_writer_args_docs)
+
+    def __init__(self, where, file_version='0.12',
+                 batch_size=1024,
+                 stripe_size=67108864,
+                 compression='zlib',
+                 compression_block_size=65536,
+                 compression_strategy='speed',
+                 row_index_stride=10000,
+                 padding_tolerance=0.0,
+                 dictionary_key_size_threshold=0.0,
+                 bloom_filter_columns=None,
+                 bloom_filter_fpp=0.05,
+                 ):
 
-    def __init__(self, where):
         self.writer = _orc.ORCWriter()
-        self.writer.open(where)
+        self.writer.open(
+            where,
+            file_version=file_version,
+            batch_size=batch_size,
+            stripe_size=stripe_size,
+            compression=compression,
+            compression_block_size=compression_block_size,
+            compression_strategy=compression_strategy,
+            row_index_stride=row_index_stride,
+            padding_tolerance=padding_tolerance,
+            dictionary_key_size_threshold=dictionary_key_size_threshold,
+            bloom_filter_columns=bloom_filter_columns,
+            bloom_filter_fpp=bloom_filter_fpp
+        )
+        self.is_open = True
+
+    def __del__(self):
+        if getattr(self, 'is_open', False):
+            self.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.close()
+        # return false since we want to propagate exceptions
```

Comment:
Isn't propagation the default?

Reply:
Ah yes. That part is inspired by existing code in parquet.